src/assign.c

```diff
@@ -529,6 +529,8 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
     else if (TRUELENGTH(names) != oldtncol)
       // Use (long long) to cast R_xlen_t to a fixed type to robustly avoid -Wformat compiler warnings, see #5768, PRId64 didn't work
       internal_error(__func__, "selfrefnames is ok but tl names [%lld] != tl [%d]", (long long)TRUELENGTH(names), oldtncol);  // # nocov
+    if (!selfrefok(dt, verbose)) // #6410 setDT(dt) and subsequent attr<- can lead to invalid selfref
```

Comment:
Change looks reasonable, one lingering question: have we gotten the root cause here? Is `attr<-` the _only_ way to reach this error? e.g. what about `class<-` or `attributes<-` or `names<-`? I don't see anything in the code nearby here that speaks specifically to attributes, so if this error can trigger in some other way, it may be confusing to the user to be admonished about `attr<-`.

Reply:
[`rlang::duplicate(shallow = TRUE)`](https://github.com/Rdatatable/data.table/pull/6419#discussion_r1735658681) may be contrived, so here's an example without it:

```r
mtcars <- data.table::setDT(data.table::copy(datasets::mtcars))
mtcars[[length(mtcars)]] <- 42
data.table::set(mtcars, NULL, paste0("test", 1:10), NA_character_)
```

The root cause is the same, a `shallow_duplicate` done by R. (Not entirely sure why.)

Reply:
Good point. I also thought about other cases but all your examples basically manipulate underlying attributes. As Ivan pointed out, the error boils down to R doing a shallow copy

Reply:
Got it. Because it's R doing the copy, I also doubt that we'll be able to intercept every possible way for this to happen, too -- checking `selrefok()` in a crucial region is the best we can do as I see it.